airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/MongoDbStateIterator.java

```diff
@@ -0,0 +1,136 @@
+ *
+ * Will also output a state message as the last message after the wrapper iterator has completed.
+ */
+class MongoDbStateIterator implements Iterator<AirbyteMessage> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoDbStateIterator.class);
+
+  private final MongoCursor<Document> iter;
+  private final ConfiguredAirbyteStream stream;
+  private final List<String> fields;
+  private final Instant emittedAt;
+  private final Integer checkpointInterval;
+
+  /**
+   * Counts the number of records seen in this batch, resets when a state-message has been generated.
+   */
+  private int count = 0;
+
+  /**
+   * Pointer to the last document _id seen by this iterator, necessary to track for state messages.
+   */
+  private String lastId = null;
+
+  /**
+   * This iterator outputs a final state when the wrapped `iter` has concluded. When this is true, the
+   * final message will be returned.
+   */
+  private boolean finalStateNext = false;
+
+  /**
+   * Constructor.
+   *
+   * @param iter MongoCursor that iterates over Mongo documents
+   * @param stream the stream that this iterator represents
+   * @param state the initial state of this stream
+   * @param emittedAt when this iterator was started
+   * @param checkpointInterval how often a state message should be emitted.
+   */
+  MongoDbStateIterator(final MongoCursor<Document> iter,
+      final ConfiguredAirbyteStream stream,
+      Optional<MongodbStreamState> state,
+      final Instant emittedAt,
+      final int checkpointInterval) {
+    this.iter = iter;
+    this.stream = stream;
+    this.checkpointInterval = checkpointInterval;
+    this.emittedAt = emittedAt;
+    fields = CatalogHelpers.getTopLevelFieldNames(stream).stream().toList();
+    lastId = state.map(MongodbStreamState::id).orElse(null);
+  }
+
+  @Override
+  public boolean hasNext() {
+    try {
+      if (iter.hasNext()) {
+        return true;
+      }
+    } catch (MongoException e) {
+      // If hasNext throws an exception, log it and then treat it as if hasNext returned false.
+      LOGGER.info("hasNext threw an exception: {}", e.getMessage(), e);
+    }
+
+    if (!finalStateNext) {
+      finalStateNext = true;
+      return true;
+    }
+
+    return false;
+  }
+
+  @Override
+  public AirbyteMessage next() {
+    if ((count > 0 && count % checkpointInterval == 0) || finalStateNext) {
+      count = 0;
+
+      final var streamState = new AirbyteStreamState()
+          .withStreamDescriptor(new StreamDescriptor()
+              .withName(stream.getStream().getName())
+              .withNamespace(stream.getStream().getNamespace()));
+      if (lastId != null) {
+        // TODO add type support in here once more than ObjectId fields are supported
+        streamState.withStreamState(Jsons.jsonNode(new MongodbStreamState(lastId)));
+      }
+
+      final var stateMessage = new AirbyteStateMessage()
+          .withType(AirbyteStateType.STREAM)
```

Comment:
If you plan to build CDC on top of this, CDC is always of GLOBAL state type and not PER STREAM

Reply:
We will adjust this once we work on the CDC state.  This is definitely one of those "in progress" things in this PR.